library/src/cmake/suitesparse/FindCHOLMOD.cmake

```diff
@@ -0,0 +1,73 @@
+# ########################################################################
+# Copyright (C) 2024 Advanced Micro Devices, Inc. All rights reserved.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell cop-
+# ies of the Software, and to permit persons to whom the Software is furnished
+# to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IM-
+# PLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+# FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+# COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+# IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNE-
+# CTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+#
+# ########################################################################
+
+#[=======================================================================[.rst:
+FindCHOLMOD
+----------
+
+Find the SuiteSparse CHOLMOD library
+
+Imported targets
+^^^^^^^^^^^^^^^^
+
+This module defines the :prop_tgt:`IMPORTED` target if CHOLMOD is found:
+
+``SuiteSparse::CHOLMOD``
+
+Result Variables
+^^^^^^^^^^^^^^^^
+
+This module defines the following variables:
+
+``CHOLMOD_INCLUDE_DIR``
+``CHOLMOD_LIBRARY``
+``CHOLMOD_LIBRARIES``
+``CHOLMOD_FOUND``
+
+#]=======================================================================]
+
+find_path(CHOLMOD_INCLUDE_DIR suitesparse/cholmod.h)
+find_library(CHOLMOD_LIBRARY cholmod)
+
+include(FindPackageHandleStandardArgs)
+find_package_handle_standard_args(CHOLMOD CHOLMOD_INCLUDE_DIR CHOLMOD_LIBRARY)
+
+find_package(SuiteSparse_config QUIET)
+
+if(CHOLMOD_FOUND)
+  if(NOT DEFINED CHOLMOD_LIBRARIES)
+    if(TARGET SuiteSparse::SuiteSparse_config)
+      set(CHOLMOD_LIBRARIES ${CHOLMOD_LIBRARY} SuiteSparse::SuiteSparse_config)
+    else()
+      set(CHOLMOD_LIBRARIES ${CHOLMOD_LIBRARY})
+    endif()
+  endif()
+
+  if(NOT TARGET SuiteSparse::CHOLMOD)
+    add_library(SuiteSparse::CHOLMOD INTERFACE IMPORTED)
```

Comment:
I think there is a typo here, as [hipsolver consumes](https://github.com/ROCm/hipSOLVER/blob/13ebb6708a855238e344ca63670d3c6aad5af893/library/src/CMakeLists.txt#L146) the target [`CHOLMOD`](https://github.com/DrTimothyAldenDavis/SuiteSparse/blob/26ababc7f3af725c5fb9168a1b94850eab74b666/CHOLMOD/CMakeLists.txt#L328) instead of the alias [`Suitesparse:CHOLMOD`](https://github.com/DrTimothyAldenDavis/SuiteSparse/blob/26ababc7f3af725c5fb9168a1b94850eab74b666/CMakeLists.txt#L360). Please, provide targets and aliases to match the ones originally defined by SuiteSparse.

Notwithstanding the previous syntactic issue, this code should work for many use cases. However, I disagree with the use of the `INTERFACE` property here. I understand that interface is meant to set requirements and define usage, and not to point to actual libraries on disk ([per the docs](https://cmake.org/cmake/help/latest/command/add_library.html#imported-libraries)). Thus, the canonical procedure would be to import those targets either as `STATIC` or `SHARED` libraries and at least set their `IMPORTED_LOCATION` (or `IMPORTED_LOCATION_<CONFIG>`) property (but you likely want to set both the `IMPORTED_LOCATION*` and `IMPORTED_IMPLIB*` properties for Windows users that might use this as well).

As is, cmake might not be able to properly handle exporting or installing hipsolver or clients that use those `Find*.cmake` files downstream. Moreover, Windows users will be able to compile hipsolver, but they might have issues running clients linking to hipsolver if the required SuiteSparse dlls are not in the path (as SuiteSparse's dlls won't be installed alongside hipsolver runtime).

Reply:
> However, I disagree with the use of the INTERFACE property here. I understand that interface is meant to set requirements and define usage, and not to point to actual libraries on disk ([per the docs](https://cmake.org/cmake/help/latest/command/add_library.html#imported-libraries)). Thus, the canonical procedure would be to import those targets either as STATIC or SHARED libraries

I was following the example of FindLAPACK, but that one is not specifically a single library. It should probably be `UNKNOWN` rather than `IMPORTED`. I'll take a look.

> As is, cmake might not be able to properly handle exporting or installing hipsolver or clients that use those Find*.cmake files downstream.

Good point.